### ISSUE TITLE
Enhancement: Refine feature deprecation process

### ISSUE THREAD

**Issue by meeech** [argoproj/argoproj#248]
People got 'caught out' by a recent feature deprecation in  Argo CD

* People weren't really aware it was happening (warnings leading up to this were missed/not visible enough)
* There turned out to be an issue, which prolly would have been caught if more people switched to using the new feature replacing the deprecated feature earlier

Suggestion: Instead of warning/removing feature, add a step where we explicitly disable the feature

Current: 
* ... 
* 2.7: Warning, this feature will be removed in 2.8
* 2.8: Feature/code removed

Proposed
* ... 
* 2.7: Warning, this feature will be removed in 2.8
* 2.8: Feature disabled. Can be re-enabled via a flag
* 2.9: Feature/code removed

This is something all projects would have to deal with, while implementation would be different for each project, the general pattern I think can be agreed on at the argoproj level.

**Comment by todaywasawesome** [comment]
From the contributor's meeting this morning. There was a bit push to start fully adopting semver. Generally major versions are the only ones meant to include breaking changes. I think this is deadon and should be our policy. 

However, I think without some additional UI tells that deprecation is happening people will still miss it.

**Comment by hblixt** [comment]
> I think I'd be willing to sacrifice the UX of those who don't read major release notes for the lower maintenance burden of avoiding a UI-based deprecation warning system.

Agreed. We should be able to assume that people at least read the deprecation warning section of the minor version release notes or the breaking changes section of the major version release notes.

**Comment by agilgur5** [comment]
I mentioned that Workflows explicitly does not follow SemVer per its docs in https://github.com/argoproj/argo-cd/pull/18111#pullrequestreview-2043809639 and [today's Maintainers Meeting](https://docs.google.com/document/d/1C71GA0XCxsyGP7Q4SYveHak5ZZcJOuaOiDCt_xHG10U/edit#heading=h.aywjjtr02vuf)

With regard to deprecation, Workflows has historically made breaking changes with at least 1 or 2 versions of backward-compat / deprecation window of the old feature. Workflows's Executors are one example of that, see also https://github.com/argoproj/argo-workflows/pull/12975

Workflows also does have a [UI popup modal](https://github.com/argoproj/argo-workflows/blob/a4fc31887041e80f7f8f51836bbf692a6ed83c98/ui/src/app/modals/new-version-modal.tsx) whenever the version changes, but popups in general are commonly ignored on the web due to their proliferation, so I can't say how effective it is. I know I've missed updates there myself in the past as I just automatically close pop-ups in my browser 😅
